File: problems/001-Two-Sum.js
/*
1. Two Sum
Given an array of integers, return indices of the two numbers such that they add up to a specific target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

Example:
Given nums = [2, 7, 11, 15], target = 9,

Because nums[0] + nums[1] = 2 + 7 = 9,
return [0, 1].

给一个里面元素为int的阵列，阵列中会有两个元素加起来等于target，回传这两个元素的位置。 +

范例：
[2, 7, 11, 15], target = 9，2+7=9，因此回传[1,2]

*/


// 需要注意的点是 nums = [3, 3] , target = 6。

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function (nums, target) {
    var ret = [];
    let index1 = 0;
    let index2 = 0;
    for (let num of nums) {
        if ((index2 = nums.lastIndexOf(target - num)) > index1) {
            ret.push(index1, index2);
            break;
        }
        index1++;
    }
    return ret;
};

module.exports = twoSum;

File: problems/001-two-sum.cpp
/*
 * [1] Two Sum
 *
 * https://leetcode-cn.com/problems/two-sum/description/
 *
 * algorithms
 * Easy (40.84%)
 * Total Accepted:    5K
 * Total Submissions: 12.1K
 * Testcase Example:  '[3,2,4]\n6'
 *
 * 给定一个整数数列，找出其中和为特定值的那两个数。
 * 你可以假设每个输入都只会有一种答案，同样的元素不能被重用。
 * 示例:
 * 给定 nums = [2, 7, 11, 15], target = 9
 * 因为 nums[0] + nums[1] = 2 + 7 = 9
 * 所以返回 [0, 1]
 */

#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
using namespace std;

class Solution
{
  public:
    vector<int> twoSum(vector<int> &nums, int target)
    {
        vector<int> ret;
        for (auto iter = nums.begin(); iter < nums.end(); iter++)
        {
            auto iFind = find(iter + 1, nums.end(), (target - *iter));
            if (iFind != nums.end())
            {
                ret.push_back(iter - nums.begin());
                ret.push_back(iFind - nums.begin());
                break;
            }
        }
        return ret;
    }
};

int main()
{
    Solution s;
    vector<int> nums{8, 11, 2, 7};
    int target = 9;

    vector<int> ret = s.twoSum(nums, target);

    for_each(ret.begin(), ret.end(), [](int &n){ cout << n << " "; });

    return 0;
}
